sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/inspections/UnusedColumnInspection.kt

```diff
@@ -0,0 +1,136 @@
+
+import com.alecstrong.sql.psi.core.psi.SqlColumnDef
+import com.alecstrong.sql.psi.core.psi.SqlCreateTableStmt
+import com.alecstrong.sql.psi.core.psi.SqlSelectStmt
+import com.alecstrong.sql.psi.core.psi.SqlTableName
+import com.alecstrong.sql.psi.core.psi.SqlTypes
+import com.alecstrong.sql.psi.core.psi.SqlVisitor
+import com.intellij.codeInspection.LocalInspectionTool
+import com.intellij.codeInspection.LocalInspectionToolSession
+import com.intellij.codeInspection.LocalQuickFixOnPsiElement
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.openapi.command.WriteCommandAction
+import com.intellij.openapi.editor.ReadOnlyModificationException
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementVisitor
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiFileFactory
+import com.intellij.psi.PsiManager
+import com.intellij.psi.SmartPointerManager
+import com.intellij.psi.search.FileTypeIndex
+import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.psi.search.searches.ReferencesSearch
+import com.intellij.psi.util.elementType
+import com.squareup.sqldelight.core.lang.SqlDelightFile
+import com.squareup.sqldelight.core.lang.SqlDelightFileType
+import com.squareup.sqldelight.core.lang.util.findChildOfType
+import com.squareup.sqldelight.core.lang.util.findChildrenOfType
+import com.squareup.sqldelight.core.lang.util.nextSibling
+import com.squareup.sqldelight.core.lang.util.prevSibling
+
+class UnusedColumnInspection : LocalInspectionTool() {
+
+  override fun buildVisitor(
+    holder: ProblemsHolder,
+    isOnTheFly: Boolean,
+    session: LocalInspectionToolSession
+  ): PsiElementVisitor = object : SqlVisitor() {
+    override fun visitCreateTableStmt(o: SqlCreateTableStmt) {
+      val candidates = o.findChildrenOfType<SqlColumnDef>()
+        .filter { columnDef ->
+          val columnName = columnDef.columnName
+          ReferencesSearch.search(columnName, columnName.useScope)
+            .allMatch { reference -> reference.element.parent is SqlColumnDef }
+        }
+
+      if (candidates.isEmpty()) {
+        return
+      }
+
+      val project = o.project
+      val psiManager = PsiManager.getInstance(project)
+      val files = FileTypeIndex.getFiles(SqlDelightFileType, GlobalSearchScope.allScope(project))
+      val hasSelectStar = files.asSequence()
+        .map { virtualFile -> psiManager.findFile(virtualFile) as SqlDelightFile }
+        .flatMap { file -> file.sqlStmtList?.stmtList.orEmpty() }
+        .filter { stmt ->
+          stmt.compoundSelectStmt != null && stmt.findChildrenOfType<SqlTableName>()
+            .any { name -> name.textMatches(o.tableName.text) }
+        }
+        .flatMap { stmt -> stmt.findChildrenOfType<SqlSelectStmt>() }
+        .flatMap { stmt -> stmt.resultColumnList }
+        .any { column -> column.textMatches("*") }
```

Comment:
I suspect this will run into issues with views and aliases, but I think we can cover both with some existing apis:

```kotlin
private fun PsiElement.columnDef(): SqlColumnDef? {
  if (this is SqlColumnDef) return this
  if (this is AliasElement) return source().columnDef()
  return null
}

FileTypeIndex.getFiles(SqlDelightFileType, GlobalSearchScope.allScope(project))
  .mapNotNull { vFile -> psiManager.findFile(vFile) as SqlDelightFile? }
  .asSequence() // Less memory intensive
  .flatMap { file -> file.sqlStmtList?.stmtList.orEmpty() }
  .flatMap { stmt -> stmt.compoundSelectStmt?.queryExposed().orEmpty() } // List<QueryResult>
  .flatMap { it.columns } // List<QueryColumn>
  .forEach { column -> column.element.columnDef()?.let(candidates::remove) }
  
candidates.forEach { ... }
```